src/transcoding.ts

```diff
@@ -257,10 +260,50 @@ export function requestChangeCaseAndCleanup(
   return convertedRequest;
 }
 
+export function isProto3OptionalField(field: Field) {
+  return field && field.options && field.options![proto3OptionalName];
+}
+
+export function isRequiredField(field: Field) {
+  return (
+    field &&
+    field.options &&
+    field.options![fieldBehaviorOptionName] === 'REQUIRED'
+  );
+}
+
+export function getFieldNameOnBehavior(
+  fields: {[k: string]: Field} | undefined
+) {
+  const requiredFields = new Set<string>();
+  const optionalFields = new Set<string>();
+  for (const fieldName in fields) {
```

Comment:
Can you do `for (const [key, value] of Object.entries(fields)` ? (Whatever's most idiomatic, just expecting it because it's possible in some other languages)

Reply:
Actually, `fieldName` is a bad name here - it's actually a `fieldIndex` :)   Yes, it's worth refactoring as @vchudnov-g suggests (I missed this during my review).  Not a big deal though.

Reply:
@summer-ji-eng FYI ↑↑↑